File: bots/applicationBot.ts
import discord from 'discord.js';
import { logger } from '../services/logger';

const { Client, GatewayIntentBits, EmbedBuilder, ActionRowBuilder, ButtonBuilder, ButtonStyle, PermissionFlagsBits } = discord;

const intents = [
  GatewayIntentBits.Guilds,
  GatewayIntentBits.GuildMessages,
  GatewayIntentBits.MessageContent,
  GatewayIntentBits.GuildMembers,
];

const bot = new Client({ intents });

const ASTDX_ROLE_ID = '1382679356975087647';
const ALS_ROLE_ID = '1397023739631108188';
const CATEGORY_ID = '1397234544368685269';
const STAFF_ROLE_ID = '1397186487044411522';

let application_counter = {
  "astdx": 1,
  "als": 1,
  "all": 1
};

const form_message = (
  "Welcome! We are glad that you want to become our important part - Helpers, but we need to clarify that you could do it or not.\n\n" +
  "**1. Respect**\n" +
  "- You must respect people, but if they don't respect or are not grateful for your help, please make a report with proofs.\n\n" +
  "**2. Choices**\n" +
  "- You can choose what requests you want to do. If you see a request you can't handle, pass it to someone else.\n\n" +
  "**3. Ratings**\n" +
  "- After finishing the request, ask them to click 'Delete Ticket' and a rating board will appear. With enough 5-star ratings, you can join special giveaways!"
);

bot.once('ready', () => {
  logger.info(`Application Bot logged in as ${bot.user?.tag}`);
});

bot.on('interactionCreate', async (interaction) => {
  if (!interaction.isButton()) return;

  const { customId, user, guild } = interaction;

  if (customId.startsWith('apply_')) {
    const appType = customId.split('_')[1];
    const category = guild?.channels.cache.get(CATEGORY_ID);

    const number = application_counter[appType as keyof typeof application_counter];
    application_counter[appType as keyof typeof application_counter] += 1;

    const channelName = `${appType}-application-${number.toString().padStart(4, '0')}`;
    
    try {
      const channel = await guild?.channels.create({
        name: channelName,
        type: discord.ChannelType.GuildText,
        parent: category?.id,
        permissionOverwrites: [
          {
            id: guild.roles.everyone.id,
            deny: [PermissionFlagsBits.ViewChannel],
          },
          {
            id: user.id,
            allow: [PermissionFlagsBits.ViewChannel, PermissionFlagsBits.SendMessages],
          },
        ],
      });

      await interaction.reply({ content: `✅ Application created: ${channel}`, ephemeral: true });

      await channel?.send(`${user}\n${form_message}`);

      const row = new ActionRowBuilder<ButtonBuilder>()
        .addComponents(
          new ButtonBuilder()
            .setCustomId(`accept_${appType}_${user.id}`)
            .setLabel('✅ Accept')
            .setStyle(ButtonStyle.Success),
          new ButtonBuilder()
            .setCustomId(`reject_${user.id}`)
            .setLabel('❌ Reject')
            .setStyle(ButtonStyle.Danger),
          new ButtonBuilder()
            .setCustomId(`delete_${channel?.id}_${user.id}`)
            .setLabel('🗑️ Delete Ticket')
            .setStyle(ButtonStyle.Danger)
        );

      await channel?.send({ content: 'Staff controls:', components: [row] });
    } catch (error) {
      logger.error('Error creating application channel:', error);
      await interaction.reply({ content: 'Failed to create application channel.', ephemeral: true });
    }
  }

  if (customId.startsWith('accept_')) {
    const parts = customId.split('_');
    const appType = parts[1];
    const userId = parts[2];

    const member = guild?.members.cache.get(userId);
    if (!member) return;

    const roles = [];
    if (appType === 'astdx') {
      roles.push(ASTDX_ROLE_ID);
    } else if (appType === 'als') {
      roles.push(ALS_ROLE_ID);
    } else if (appType === 'all') {
      roles.push(ALS_ROLE_ID, ASTDX_ROLE_ID);
    }

    try {
      for (const roleId of roles) {
        const role = guild?.roles.cache.get(roleId);
        if (role) await member.roles.add(role);
      }

      await interaction.reply({ content: `✅ ${member} has been accepted and given roles.`, ephemeral: true });
    } catch (error) {
      logger.error('Error adding roles:', error);
    }
  }

  if (customId.startsWith('reject_')) {
    const userId = customId.split('_')[1];
    const member = guild?.members.cache.get(userId);
    await interaction.reply({ content: `❌ ${member}'s application was rejected.`, ephemeral: true });
  }

  if (customId.startsWith('delete_')) {
    const channelId = customId.split('_')[1];
    const channel = guild?.channels.cache.get(channelId);
    
    await interaction.reply({ content: 'This ticket will be deleted in 5 seconds...', ephemeral: true });
    setTimeout(async () => {
      try {
        await channel?.delete();
      } catch (error) {
        logger.error('Error deleting channel:', error);
      }
    }, 5000);
  }
});

// Application command
bot.on('messageCreate', async (message) => {
  if (message.author.bot) return;
  if (!message.content.startsWith('app!')) return;

  const args = message.content.slice(4).trim().split(/ +/);
  const command = args.shift()?.toLowerCase();

  if (command === 'application') {
    const row = new ActionRowBuilder<ButtonBuilder>()
      .addComponents(
        new ButtonBuilder()
          .setCustomId('apply_astdx')
          .setLabel('Apply on ASTDX')
          .setStyle(ButtonStyle.Primary),
        new ButtonBuilder()
          .setCustomId('apply_als')
          .setLabel('Apply on ALS')
          .setStyle(ButtonStyle.Primary),
        new ButtonBuilder()
          .setCustomId('apply_all')
          .setLabel('Apply All')
          .setStyle(ButtonStyle.Primary)
      );

    await message.channel.send({ content: 'Click a button below to apply:', components: [row] });
  }

  if (command === 'rules') {
    const embed1 = new EmbedBuilder()
      .setTitle('📜 AVS Server Rules & Guidelines')
      .setDescription(
        "**Welcome to AVS — a fun community for ALS, ASTDX, and chill chats!**\n\n" +
        "**🌟 1. Respect & Kindness**\n" +
        "• Be respectful to everyone.\n" +
        "• No harassment, hate, threats, or toxic behavior — jokes are fine, but not harmful ones.\n" +
        "• If someone bothers you, don't retaliate — report them.\n\n" +
        "**📢 2. Language & Behavior**\n" +
        "• Keep chats appropriate for all ages.\n" +
        "• Swearing is okay in moderation — no slurs or hate speech.\n" +
        "• No spam, flooding, mass pings.\n\n" +
        "**📌 3. Use the Right Channels**\n" +
        "• Chat in general, memes in #memes, help in #support, etc.\n\n" +
        "**🧠 4. Content Rules**\n" +
        "• No NSFW/NSFL, illegal, pirated, or disturbing content.\n\n" +
        "**🚫 5. Alts & Raids**\n" +
        "• No alts unless approved. Trolls = instant ban.\n\n" +
        "**✅ Final:**\n" +
        "• No disrespect to helpers. Major = ban. Minor = mute.\n"
      )
      .setColor(0x0099ff);

    const nextButton = new ActionRowBuilder<ButtonBuilder>()
      .addComponents(
        new ButtonBuilder()
          .setCustomId(`rules_next_${message.author.id}`)
          .setLabel('Next ▶️')
          .setStyle(ButtonStyle.Primary),
        new ButtonBuilder()
          .setCustomId(`rules_back_${message.author.id}`)
          .setLabel('◀️ Back')
          .setStyle(ButtonStyle.Secondary)
          .setDisabled(true)
      );

    await message.channel.send({ embeds: [embed1], components: [nextButton] });
  }
});

// Handle rules pagination
bot.on('interactionCreate', async (interaction) => {
  if (!interaction.isButton()) return;

  const { customId, user } = interaction;

  if (customId.startsWith('rules_next_')) {
    const userId = customId.split('_')[2];
    if (user.id !== userId) {
      await interaction.reply({ content: '🚫 You don\'t have permission to use this button.', ephemeral: true });
      return;
    }

    const member = interaction.guild?.members.cache.get(userId);
    const staffRole = interaction.guild?.roles.cache.get(STAFF_ROLE_ID);
    
    if (staffRole && member?.roles.cache.has(STAFF_ROLE_ID)) {
      const embed2 = new EmbedBuilder()
        .setTitle('👮 Staff Guide')
        .setDescription(
          "• **Take Requests:** When a ticket appears, check what the member needs. If you're confident you can help, press **TAKE REQUEST**.\n\n" +
          "• **Deletion & Ratings:** After completing the request, ask the member to press **DELETE TICKET** to leave a rating.\n\n" +
          "• **Choices:** You don't have to take every ticket. Only take the ones you can handle.\n\n" +
          "• **Respect:** Always respect members. If someone disrespects you, stop helping and report it.\n\n" +
          "• **Benefits:** Staff can join exclusive giveaways and perks."
        )
        .setColor(0xff0000);

      const backButton = new ActionRowBuilder<ButtonBuilder>()
        .addComponents(
          new ButtonBuilder()
            .setCustomId(`rules_next_${userId}`)
            .setLabel('Next ▶️')
            .setStyle(ButtonStyle.Primary)
            .setDisabled(true),
          new ButtonBuilder()
            .setCustomId(`rules_back_${userId}`)
            .setLabel('◀️ Back')
            .setStyle(ButtonStyle.Secondary)
        );

      await interaction.update({ embeds: [embed2], components: [backButton] });
    } else {
      await interaction.reply({ content: '🚫 You don\'t have permission to view this page.', ephemeral: true });
    }
  }

  if (customId.startsWith('rules_back_')) {
    const userId = customId.split('_')[2];
    if (user.id !== userId) return;

    const embed1 = new EmbedBuilder()
      .setTitle('📜 AVS Server Rules & Guidelines')
      .setDescription(
        "**Welcome to AVS — a fun community for ALS, ASTDX, and chill chats!**\n\n" +
        "**🌟 1. Respect & Kindness**\n" +
        "• Be respectful to everyone.\n" +
        "• No harassment, hate, threats, or toxic behavior — jokes are fine, but not harmful ones.\n" +
        "• If someone bothers you, don't retaliate — report them.\n\n" +
        "**📢 2. Language & Behavior**\n" +
        "• Keep chats appropriate for all ages.\n" +
        "• Swearing is okay in moderation — no slurs or hate speech.\n" +
        "• No spam, flooding, mass pings.\n\n" +
        "**📌 3. Use the Right Channels**\n" +
        "• Chat in general, memes in #memes, help in #support, etc.\n\n" +
        "**🧠 4. Content Rules**\n" +
        "• No NSFW/NSFL, illegal, pirated, or disturbing content.\n\n" +
        "**🚫 5. Alts & Raids**\n" +
        "• No alts unless approved. Trolls = instant ban.\n\n" +
        "**✅ Final:**\n" +
        "• No disrespect to helpers. Major = ban. Minor = mute.\n"
      )
      .setColor(0x0099ff);

    const nextButton = new ActionRowBuilder<ButtonBuilder>()
      .addComponents(
        new ButtonBuilder()
          .setCustomId(`rules_next_${userId}`)
          .setLabel('Next ▶️')
          .setStyle(ButtonStyle.Primary),
        new ButtonBuilder()
          .setCustomId(`rules_back_${userId}`)
          .setLabel('◀️ Back')
          .setStyle(ButtonStyle.Secondary)
          .setDisabled(true)
      );

    await interaction.update({ embeds: [embed1], components: [nextButton] });
  }
});

bot.on('error', (error) => {
  logger.error('Application Bot error:', error);
});

bot.on('disconnect', () => {
  logger.warn('Application Bot disconnected');
});

const token = process.env.BOT_TOKEN || process.env.APPLICATION_BOT_TOKEN;
if (!token) {
  logger.error('No bot token provided for Application Bot');
  process.exit(1);
}

bot.login(token).catch((error) => {
  logger.error('Failed to login Application Bot:', error);
  process.exit(1);
});

File: Application-bot.py
import discord
from discord.ext import commands
from discord.ui import Button, View
import asyncio


intents = discord.Intents.all()
bot = commands.Bot(command_prefix="app!", intents=intents)


ASTDX_ROLE_ID = 1382679356975087647
ALS_ROLE_ID = 1397023739631108188
CATEGORY_ID = 1397234544368685269
STAFF_ROLE_ID = 1397186487044411522


application_counter = {
    "astdx": 1,
    "als": 1,
    "all": 1
}


form_message = (
    "Welcome! We are glad that you want to become our important part - Helpers, but we need to clarify that you could do it or not.\n\n"
    "**1. Respect**\n"
    "- You must respect people, but if they don't respect or are not grateful for your help, please make a report with proofs.\n\n"
    "**2. Choices**\n"
    "- You can choose what requests you want to do. If you see a request you can't handle, pass it to someone else.\n\n"
    "**3. Ratings**\n"
    "- After finishing the request, ask them to click 'Delete Ticket' and a rating board will appear. With enough 5-star ratings, you can join special giveaways!"
)


class ApplicationView(View):
    def __init__(self):
        super().__init__(timeout=None)
        self.add_item(ApplicationButton("Apply on ASTDX", "astdx"))
        self.add_item(ApplicationButton("Apply on ALS", "als"))
        self.add_item(ApplicationButton("Apply All", "all"))


class ApplicationButton(Button):
    def __init__(self, label, app_type):
        super().__init__(label=label, style=discord.ButtonStyle.primary)
        self.app_type = app_type


    async def callback(self, interaction: discord.Interaction):
        user = interaction.user
        guild = interaction.guild
        category = guild.get_channel(CATEGORY_ID)


        number = application_counter[self.app_type]
        application_counter[self.app_type] += 1


        channel_name = f"{self.app_type}-application-#{number:04}"
        overwrites = {
            guild.default_role: discord.PermissionOverwrite(read_messages=False),
            user: discord.PermissionOverwrite(read_messages=True, send_messages=True),
        }


        channel = await guild.create_text_channel(
            name=channel_name,
            category=category,
            overwrites=overwrites
        )


        await interaction.response.send_message(f"✅ Application created: {channel.mention}", ephemeral=True)


        await channel.send(f"{user.mention}\n{form_message}")


        view = View()
        view.add_item(AcceptButton(self.app_type, user))
        view.add_item(RejectButton(user))
        view.add_item(DeleteTicketButton(channel, user))
        await channel.send("Staff controls:", view=view)


class AcceptButton(Button):
    def __init__(self, app_type, user):
        super().__init__(label="✅ Accept", style=discord.ButtonStyle.success)
        self.app_type = app_type
        self.user = user


    async def callback(self, interaction: discord.Interaction):
        roles = []
        if self.app_type == "astdx":
            roles.append(ASTDX_ROLE_ID)
        elif self.app_type == "als":
            roles.append(ALS_ROLE_ID)
        elif self.app_type == "all":
            roles.extend([ALS_ROLE_ID, ASTDX_ROLE_ID])


        for role_id in roles:
            role = interaction.guild.get_role(role_id)
            if role:
                await self.user.add_roles(role)


        await interaction.response.send_message(f"✅ {self.user.mention} has been accepted and given roles.", ephemeral=True)


class RejectButton(Button):
    def __init__(self, user):
        super().__init__(label="❌ Reject", style=discord.ButtonStyle.danger)
        self.user = user


    async def callback(self, interaction: discord.Interaction):
        await interaction.response.send_message(f"❌ {self.user.mention}'s application was rejected.", ephemeral=True)


class DeleteTicketButton(Button):
    def __init__(self, channel, user):
        super().__init__(label="🗑️ Delete Ticket", style=discord.ButtonStyle.danger)
        self.channel = channel
        self.user = user


    async def callback(self, interaction: discord.Interaction):
        await interaction.response.send_message("This ticket will be deleted in 5 seconds...", ephemeral=True)
        await asyncio.sleep(5)
        await self.channel.delete()


# ======================= RULES COMMAND ========================


class RulesView(View):
    def __init__(self, author: discord.Member):
        super().__init__(timeout=None)
        self.author = author
        self.current_page = 1


    async def interaction_check(self, interaction: discord.Interaction) -> bool:
        return interaction.user.id == self.author.id


    @discord.ui.button(label="Next ▶️", style=discord.ButtonStyle.primary)
    async def next_page(self, interaction: discord.Interaction, button: Button):
        if self.current_page == 1:
            staff_role = interaction.guild.get_role(STAFF_ROLE_ID)
            if staff_role in interaction.user.roles:
                embed2 = discord.Embed(
                    title="👮 Staff Guide",
                    description=(
                        "• **Take Requests:** When a ticket appears, check what the member needs. If you're confident you can help, press **TAKE REQUEST**.\n\n"
                        "• **Deletion & Ratings:** After completing the request, ask the member to press **DELETE TICKET** to leave a rating.\n\n"
                        "• **Choices:** You don’t have to take every ticket. Only take the ones you can handle.\n\n"
                        "• **Respect:** Always respect members. If someone disrespects you, stop helping and report it.\n\n"
                        "• **Benefits:** Staff can join exclusive giveaways and perks."
                    ),
                    color=discord.Color.red()
                )
                self.current_page = 2
                await interaction.response.edit_message(embed=embed2, view=self)
            else:
                await interaction.response.send_message("🚫 You don't have permission to view this page.", ephemeral=True)


    @discord.ui.button(label="◀️ Back", style=discord.ButtonStyle.secondary)
    async def back_page(self, interaction: discord.Interaction, button: Button):
        if self.current_page == 2:
            embed1 = discord.Embed(
                title="📜 AVS Server Rules & Guidelines",
                description=(
                    "**Welcome to AVS — a fun community for ALS, ASTDX, and chill chats!**\n\n"
                    "**🌟 1. Respect & Kindness**\n"
                    "• Be respectful to everyone.\n"
                    "• No harassment, hate, threats, or toxic behavior — jokes are fine, but not harmful ones.\n"
                    "• If someone bothers you, don’t retaliate — report them.\n\n"
                    "**📢 2. Language & Behavior**\n"
                    "• Keep chats appropriate for all ages.\n"
                    "• Swearing is okay in moderation — no slurs or hate speech.\n"
                    "• No spam, flooding, mass pings.\n\n"
                    "**📌 3. Use the Right Channels**\n"
                    "• Chat in general, memes in #memes, help in #support, etc.\n\n"
                    "**🧠 4. Content Rules**\n"
                    "• No NSFW/NSFL, illegal, pirated, or disturbing content.\n\n"
                    "**🚫 5. Alts & Raids**\n"
                    "• No alts unless approved. Trolls = instant ban.\n\n"
                    "**✅ Final:**\n"
                    "• No disrespect to helpers. Major = ban. Minor = mute.\n"
                ),
                color=discord.Color.blue()
            )
            self.current_page = 1
            await interaction.response.edit_message(embed=embed1, view=self)


@bot.command()
async def rules(ctx):
    embed1 = discord.Embed(
        title="📜 AVS Server Rules & Guidelines",
        description=(
            "**Welcome to AVS — a fun community for ALS, ASTDX, and chill chats!**\n\n"
            "**🌟 1. Respect & Kindness**\n"
            "• Be respectful to everyone.\n"
            "• No harassment, hate, threats, or toxic behavior — jokes are fine, but not harmful ones.\n"
            "• If someone bothers you, don’t retaliate — report them.\n\n"
            "**📢 2. Language & Behavior**\n"
            "• Keep chats appropriate for all ages.\n"
            "• Swearing is okay in moderation — no slurs or hate speech.\n"
            "• No spam, flooding, mass pings.\n\n"
            "**📌 3. Use the Right Channels**\n"
            "• Chat in general, memes in #memes, help in #support, etc.\n\n"
            "**🧠 4. Content Rules**\n"
            "• No NSFW/NSFL, illegal, pirated, or disturbing content.\n\n"
            "**🚫 5. Alts & Raids**\n"
            "• No alts unless approved. Trolls = instant ban.\n\n"
            "**✅ Final:**\n"
            "• No disrespect to helpers. Major = ban. Minor = mute.\n"
        ),
        color=discord.Color.blue()
    )
    await ctx.send(embed=embed1, view=RulesView(ctx.author))


# ====================== APPLICATION BUTTON TRIGGER =======================


@bot.command()
async def application(ctx):
    await ctx.send("Click a button below to apply:", 
view=ApplicationView())
bot.run("MTM5NjQ5ODk2Mjg2OTEyOTM5Nw.GZpaeH.lEGyvWHhO1jpnF-mlErm07iZvBKVsFOpPyV2NY")